File: solutions/typescript/bowling/1/bowling.ts
export class Bowling {
  private rolls: number[] = [];
  private currentFrameTries = 0;
  private currentFramePins = 0;
  private frameCount = 1;

  public roll(pins: number): void {
    // 1. Validaciones básicas con los mensajes exactos del test
    if (pins < 0) {
      throw new Error('Negative roll is invalid');
    }
    if (pins > 10) {
      throw new Error('Pin count exceeds pins on the lane');
    }
    if (this.isGameOver()) {
      throw new Error('Cannot roll after game is over');
    }

    // 2. Validar que no se derriben más de 10 pines en un mismo frame (Frames 1 al 9)
    if (this.frameCount < 10) {
      if (this.currentFrameTries === 0) {
        if (pins === 10) {
          // Es un Strike: avanzamos de frame inmediatamente
          this.rolls.push(pins);
          this.frameCount++;
        } else {
          this.currentFrameTries = 1;
          this.currentFramePins = pins;
          this.rolls.push(pins);
        }
      } else {
        if (this.currentFramePins + pins > 10) {
          throw new Error('Pin count exceeds pins on the lane');
        }
        this.rolls.push(pins);
        this.currentFrameTries = 0;
        this.currentFramePins = 0;
        this.frameCount++;
      }
    } else {
      // 3. Validaciones especiales para el Frame 10 (Manejo de bolas de relleno)
      const frame10RollsCount = this.rolls.length - this.getRollsBeforeFrame10();
      
      if (frame10RollsCount === 1 && this.rolls[this.rolls.length - 1] < 10) {
        if (this.rolls[this.rolls.length - 1] + pins > 10) {
          throw new Error('Pin count exceeds pins on the lane');
        }
      } else if (frame10RollsCount === 2) {
        const firstIn10 = this.rolls[this.getRollsBeforeFrame10()];
        const secondIn10 = this.rolls[this.getRollsBeforeFrame10() + 1];
        if (firstIn10 === 10 && secondIn10 < 10 && secondIn10 + pins > 10) {
          throw new Error('Pin count exceeds pins on the lane');
        }
      }
      
      this.rolls.push(pins);
    }
  }

  public score(): number {
    if (!this.isGameOver()) {
      throw new Error('Score cannot be taken until the end of the game');
    }

    let totalScore = 0;
    let rollIndex = 0;

    for (let frame = 0; frame < 10; frame++) {
      if (this.isStrike(rollIndex)) {
        totalScore += 10 + this.rolls[rollIndex + 1] + this.rolls[rollIndex + 2];
        rollIndex += 1;
      } else if (this.isSpare(rollIndex)) {
        totalScore += 10 + this.rolls[rollIndex + 2];
        rollIndex += 2;
      } else {
        totalScore += this.rolls[rollIndex] + this.rolls[rollIndex + 1];
        rollIndex += 2;
      }
    }

    return totalScore;
  }

  // --- Métodos Auxiliares de Estado ---

  private isStrike(rollIndex: number): boolean {
    return this.rolls[rollIndex] === 10;
  }

  private isSpare(rollIndex: number): boolean {
    return this.rolls[rollIndex] + this.rolls[rollIndex + 1] === 10;
  }

  private getRollsBeforeFrame10(): number {
    let rollIndex = 0;
    for (let frame = 0; frame < 9; frame++) {
      if (this.rolls[rollIndex] === 10) {
        rollIndex += 1;
      } else {
        rollIndex += 2;
      }
    }
    return rollIndex;
  }

  private isGameOver(): boolean {
    if (this.frameCount < 10) return false;
    
    const startOfFrame10 = this.getRollsBeforeFrame10();
    const frame10Rolls = this.rolls.length - startOfFrame10;

    if (frame10Rolls < 2) return false;

    const firstIn10 = this.rolls[startOfFrame10];
    const secondIn10 = this.rolls[startOfFrame10 + 1];

    if (firstIn10 === 10 || firstIn10 + secondIn10 === 10) {
      return frame10Rolls === 3;
    }

    return frame10Rolls === 2;
  }
}

File: solutions/ruby/bowling/1/bowling.rb
class Game
  class BowlingError < StandardError; end

  def initialize
    @rolls = []
  end

  def roll(pins)
    raise BowlingError, 'Pins must have a value from 0 to 10' unless pins.between?(0, 10)
    raise BowlingError, 'Should not be able to roll after game is over' if game_over?

    # Clonamos momentáneamente para simular cómo quedaría el frame actual
    simulated_rolls = @rolls + [pins]
    frames = build_frames(simulated_rolls)
    
    # Validamos las reglas de los primeros 9 frames
    if frames.size <= 9
      current_frame = frames.last
      if current_frame && current_frame.size == 2 && current_frame.sum > 10
        raise BowlingError, 'Pin count exceeds pins on the lane'
      end
    # Validamos las reglas del décimo frame (incluyendo fill balls)
    elsif frames.size == 10
      tenth_frame = frames.last
      
      # Si los dos primeros tiros no suman un strike/spare, no pueden pasar de 10
      if tenth_frame.size == 2 && tenth_frame[0] != 10 && tenth_frame.sum > 10
        raise BowlingError, 'Pin count exceeds pins on the lane'
      end
      
      # Si el primero es strike pero el segundo no, el segundo y el tercero no pueden pasar de 10
      if tenth_frame.size == 3 && tenth_frame[0] == 10 && tenth_frame[1] != 10
        if tenth_frame[1] + tenth_frame[2] > 10
          raise BowlingError, 'Pin count exceeds pins on the lane'
        end
      end
    end

    @rolls << pins
  end

  def score
    raise BowlingError, 'Score cannot be taken until the end of the game' unless game_over?

    total_score = 0
    roll_index = 0

    10.times do
      if strike?(roll_index)
        total_score += 10 + strike_bonus(roll_index)
        roll_index += 1
      elsif spare?(roll_index)
        total_score += 10 + spare_bonus(roll_index)
        roll_index += 2
      else
        total_score += sum_of_balls_in_frame(roll_index)
        roll_index += 2
      end
    end

    total_score
  end

  private

  # Agrupa un arreglo plano de tiros en sus respectivos frames (1 al 10)
  def build_frames(rolls_to_group)
    frames = []
    current_rolls = rolls_to_group.dup

    while !current_rolls.empty? && frames.size < 9
      if current_rolls.first == 10
        frames << [current_rolls.shift] # Strike directo termina el frame
      else
        frames << current_rolls.shift(2) # Toma un par de tiros para el frame abierto/spare
      end
    end

    # Todo lo que sobre pertenece exclusivamente al décimo frame
    frames << current_rolls unless current_rolls.empty?
    frames
  end

  def game_over?
    frames = build_frames(@rolls)
    return false if frames.size < 10
    
    tenth_frame = frames[9]
    # Si el décimo frame empezó con Strike o sumó un Spare, requiere obligatoriamente 3 tiros
    if tenth_frame[0] == 10 || tenth_frame[0].to_i + tenth_frame[1].to_i == 10
      tenth_frame.size == 3
    else
      tenth_frame.size == 2
    end
  end

  def strike?(roll_index)
    @rolls[roll_index] == 10
  end

  def spare?(roll_index)
    @rolls[roll_index].to_i + @rolls[roll_index + 1].to_i == 10
  end

  def strike_bonus(roll_index)
    @rolls[roll_index + 1].to_i + @rolls[roll_index + 2].to_i
  end

  def spare_bonus(roll_index)
    @rolls[roll_index + 2].to_i
  end

  def sum_of_balls_in_frame(roll_index)
    @rolls[roll_index].to_i + @rolls[roll_index + 1].to_i
  end
end